File: src/main/java8/org/apache/beam/examples/complete/game/solutions/Exercise7.java
/*
 * Copyright (C) 2015 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package org.apache.beam.examples.complete.game.solutions;

import static org.apache.beam.sdk.transforms.windowing.TimestampCombiner.END_OF_WINDOW;

import com.google.api.services.bigquery.model.TableFieldSchema;
import com.google.api.services.bigquery.model.TableReference;
import com.google.api.services.bigquery.model.TableRow;
import com.google.api.services.bigquery.model.TableSchema;
import java.util.ArrayList;
import java.util.List;
import org.apache.beam.examples.complete.game.utils.GameEvent;
import org.apache.beam.examples.complete.game.utils.Options;
import org.apache.beam.examples.complete.game.utils.ParseEventFn;
import org.apache.beam.examples.complete.game.utils.ParsePlayEventFn;
import org.apache.beam.examples.complete.game.utils.PlayEvent;
import org.apache.beam.runners.dataflow.DataflowRunner;
import org.apache.beam.sdk.Pipeline;
import org.apache.beam.sdk.PipelineResult;
import org.apache.beam.sdk.extensions.gcp.options.GcpOptions;
import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO;
import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.Write.CreateDisposition;
import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.Write.WriteDisposition;
import org.apache.beam.sdk.io.gcp.pubsub.PubsubIO;
import org.apache.beam.sdk.metrics.Counter;
import org.apache.beam.sdk.metrics.Metrics;
import org.apache.beam.sdk.options.Description;
import org.apache.beam.sdk.options.PipelineOptionsFactory;
import org.apache.beam.sdk.options.StreamingOptions;
import org.apache.beam.sdk.options.Validation;
import org.apache.beam.sdk.transforms.ApproximateQuantiles;
import org.apache.beam.sdk.transforms.Combine;
import org.apache.beam.sdk.transforms.DoFn;
import org.apache.beam.sdk.transforms.GroupByKey;
import org.apache.beam.sdk.transforms.ParDo;
import org.apache.beam.sdk.transforms.Values;
import org.apache.beam.sdk.transforms.WithKeys;
import org.apache.beam.sdk.transforms.join.CoGbkResult;
import org.apache.beam.sdk.transforms.join.CoGroupByKey;
import org.apache.beam.sdk.transforms.join.KeyedPCollectionTuple;
import org.apache.beam.sdk.transforms.windowing.AfterPane;
import org.apache.beam.sdk.transforms.windowing.AfterProcessingTime;
import org.apache.beam.sdk.transforms.windowing.GlobalWindows;
import org.apache.beam.sdk.transforms.windowing.Repeatedly;
import org.apache.beam.sdk.transforms.windowing.Sessions;
import org.apache.beam.sdk.transforms.windowing.Window;
import org.apache.beam.sdk.values.KV;
import org.apache.beam.sdk.values.PCollection;
import org.apache.beam.sdk.values.PCollectionView;
import org.apache.beam.sdk.values.TupleTag;
import org.apache.beam.sdk.values.TypeDescriptor;
import org.joda.time.Duration;
import org.joda.time.Instant;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Seventh in a series of coding exercises in a gaming domain.
 *
 * <p>This exercise introduces concepts of pubsub message ids, session windows, side inputs, joins,
 * and global combine.
 *
 * <p>See README.md for details.
 */
public class Exercise7 {

  private static final String TIMESTAMP_ATTRIBUTE = "timestamp_ms";
  private static final String MESSAGE_ID_ATTRIBUTE = "unique_id";
  private static final int SESSION_GAP_MINUTES = 1;
  private static final int GLOBAL_LATENCY_QUANTILES = 31;
  private static final int GLOBAL_AGGREGATE_FANOUT = 16;
  private static final Logger LOG = LoggerFactory.getLogger(Exercise7.class);

  private static final TupleTag<PlayEvent> playTag = new TupleTag<PlayEvent>();
  private static final TupleTag<GameEvent> eventTag = new TupleTag<GameEvent>();

  /**
   * Options supported by {@link Exercise7}.
   */
  interface Exercise7Options extends Options, StreamingOptions {

    @Description("Pub/Sub topic to read from")
    @Validation.Required
    String getTopic();

    void setTopic(String value);

    @Description("Pub/Sub play events topic to read from")
    @Validation.Required
    String getPlayEventsTopic();

    void setPlayEventsTopic(String value);
  }

  public static class ComputeLatencyFn extends DoFn<KV<String, CoGbkResult>, KV<String, Long>> {

    private final Counter numDroppedSessionsNoEvent = Metrics
                            .counter("main", "DroppedSessionsNoEvent");
    private final Counter numDroppedSessionsTooManyEvents = Metrics
        .counter("main", "DroppedSessionsTooManyEvents");
    private final Counter numDroppedSessionsNoPlayEvents = Metrics
        .counter("main", "DroppedSessionsNoPlayEvents");

    @ProcessElement
    public void processElement(ProcessContext c) {
      Iterable<PlayEvent> plays = c.element().getValue().getAll(playTag);
      Iterable<GameEvent> events = c.element().getValue().getAll(eventTag);
      int playCount = 0;
      Long maxPlayEventTs = 0L;
      for (PlayEvent play : plays) {
        playCount++;
        maxPlayEventTs = Math.max(maxPlayEventTs, play.getTimestamp());
      }
      int eventCount = 0;
      GameEvent event = null;
      for (GameEvent currentEvent : events) {
        eventCount++;
        event = currentEvent;
      }
      String id = c.element().getKey();
      if (eventCount == 0) {
        numDroppedSessionsNoEvent.inc();
      } else if (eventCount > 1) {
        numDroppedSessionsTooManyEvents.inc();
      } else if (playCount == 0) {
        numDroppedSessionsNoPlayEvents.inc();
      } else {
        Long minLatency = event.getTimestamp() - maxPlayEventTs;
        c.output(KV.of(event.getUser(), minLatency));
      }
    }
  }

  public static void main(String[] args) throws Exception {
    Exercise7Options options =
        PipelineOptionsFactory.fromArgs(args).withValidation().as(Exercise7Options.class);
    // Enforce that this pipeline is always run in streaming mode.
    options.setStreaming(true);
    options.setRunner(DataflowRunner.class);
    Pipeline pipeline = Pipeline.create(options);

    TableReference badUserTable = new TableReference();
    badUserTable.setDatasetId(options.getOutputDataset());
    badUserTable.setProjectId(options.as(GcpOptions.class).getProject());
    badUserTable.setTableId(options.getOutputTableName() + "_bad_users");

    // Read Events from Pub/Sub using custom timestamps and custom message id label.
    PCollection<KV<String, GameEvent>> sessionedEvents =
        pipeline
            .apply(
                "ReadGameScoreEvents",
                PubsubIO.readStrings().withTimestampAttribute(TIMESTAMP_ATTRIBUTE)
                    .withIdAttribute(MESSAGE_ID_ATTRIBUTE).fromTopic(options.getTopic()))
            .apply("ParseGameScoreEvents", ParDo.of(new ParseEventFn()))
            .apply(
                "KeyGameScoreByEventId",
                WithKeys.of((GameEvent event) -> event.getEventId())
                    .withKeyType(TypeDescriptor.of(String.class)))
            .apply(
                "SessionizeGameScoreEvents",
                Window.<KV<String, GameEvent>>into(
                    Sessions.withGapDuration(Duration.standardMinutes(SESSION_GAP_MINUTES)))
                    .withTimestampCombiner(END_OF_WINDOW));

    // Read PlayEvents from Pub/Sub using custom timestamps and custom message id label.
    PCollection<KV<String, PlayEvent>> sessionedPlayEvents =
        pipeline
            .apply(
                "ReadGamePlayEvents",
                PubsubIO.readStrings().withTimestampAttribute(TIMESTAMP_ATTRIBUTE)
                    .withIdAttribute(MESSAGE_ID_ATTRIBUTE)
                    .fromTopic(options.getPlayEventsTopic()))
            .apply("ParseGamePlayEvents", ParDo.of(new ParsePlayEventFn()))
            .apply(
                "KeyGamePlayByEventId",
                WithKeys.of((PlayEvent play) -> play.getEventId())
                    .withKeyType(TypeDescriptor.of(String.class)))
            .apply(
                "SessionizeGamePlayEvents",
                Window.<KV<String, PlayEvent>>into(
                    Sessions.withGapDuration(Duration.standardMinutes(SESSION_GAP_MINUTES)))
                    .withTimestampCombiner(END_OF_WINDOW));

    // Compute per-user latency.
    PCollection<KV<String, Long>> userLatency =
        KeyedPCollectionTuple.of(playTag, sessionedPlayEvents)
            .and(eventTag, sessionedEvents)
            .apply("JoinScorePlayEvents", CoGroupByKey.create())
            .apply("ComputeLatency", ParDo.of(new ComputeLatencyFn()));

    // Create a view onto quantiles of the global latency distribution.
    PCollectionView<List<Long>> globalQuantiles =
        userLatency
            .apply("GetLatencies", Values.create())
            // Re-window session results into a global window, and trigger periodically making sure
            // to use the full accumulated window contents.
            .apply(
                "GlobalWindowRetrigger",
                Window.<Long>into(new GlobalWindows())
                    .triggering(Repeatedly.forever(AfterPane.elementCountAtLeast(1000)))
                    .accumulatingFiredPanes())
            .apply(
                ((Combine.Globally<Long, List<Long>>)
                    ApproximateQuantiles.<Long>globally(GLOBAL_LATENCY_QUANTILES))
                    .withFanout(GLOBAL_AGGREGATE_FANOUT)
                    .asSingletonView());

    userLatency
        // Use the computed latency distribution as a side-input to filter out likely bad users.
        .apply(
            "DetectBadUsers",
            ParDo
                .of(
                    new DoFn<KV<String, Long>, String>() {
                      @ProcessElement
                      public void processElement(ProcessContext c) {
                        String user = c.element().getKey();
                        Long latency = c.element().getValue();
                        List<Long> quantiles = c.sideInput(globalQuantiles);
                        // Users in the first quantile are considered spammers, since their
                        // score to play event latency is too low, suggesting a robot.
                        if (latency < quantiles.get(1)) {
                          c.output(user);
                        }
                      }
                    }).withSideInputs(globalQuantiles))
        // We want to only emit a single BigQuery row for every bad user. To do this, we
        // re-key by user, then window globally and trigger on the first element for each key.
        .apply(
            "KeyByUser",
            WithKeys.of((String user) -> user).withKeyType(TypeDescriptor.of(String.class)))
        .apply(
            "GlobalWindowsTriggerOnFirst",
            Window.<KV<String, String>>into(new GlobalWindows())
                .triggering(
                    AfterProcessingTime.pastFirstElementInPane()
                        .plusDelayOf(Duration.standardSeconds(10)))
                .accumulatingFiredPanes())
        .apply("GroupByUser", GroupByKey.<String, String>create())
        .apply("FormatBadUsers", ParDo.of(new FormatBadUserFn()))
        .apply(
            "WriteBadUsers",
            BigQueryIO.writeTableRows().to(badUserTable)
                .withSchema(FormatBadUserFn.getSchema())
                .withCreateDisposition(CreateDisposition.CREATE_IF_NEEDED)
                .withWriteDisposition(WriteDisposition.WRITE_APPEND));

    PipelineResult result = pipeline.run();
    result.waitUntilFinish();
  }

  /**
   * Format a KV of user and associated properties to a BigQuery TableRow.
   */
  static class FormatBadUserFn extends DoFn<KV<String, Iterable<String>>, TableRow> {

    @ProcessElement
    public void processElement(ProcessContext c) {
      TableRow row =
          new TableRow()
              .set("bad_user", c.element().getKey())
              .set("time", Instant.now().getMillis() / 1000);
      c.output(row);
    }

    static TableSchema getSchema() {
      List<TableFieldSchema> fields = new ArrayList<>();
      fields.add(new TableFieldSchema().setName("bad_user").setType("STRING"));
      fields.add(new TableFieldSchema().setName("time").setType("TIMESTAMP"));
      return new TableSchema().setFields(fields);
    }
  }
}

File: py/exercises/exercise7.py
from __future__ import absolute_import

import logging
import re
import time

import apache_beam as beam
import apache_beam.transforms.window as window
from apache_beam.io import ReadFromText
from apache_beam.io import ReadFromPubSub
from apache_beam.io import WriteToText
from apache_beam.metrics import Metrics
from apache_beam.metrics.metric import MetricsFilter
from apache_beam.options.pipeline_options import PipelineOptions
from apache_beam.options.pipeline_options import SetupOptions
from apache_beam.options.pipeline_options import StandardOptions
from apache_beam.options.pipeline_options import GoogleCloudOptions
from apache_beam.transforms import trigger
from util.util import GameEvent
from util.util import ParseEvent
from util.util import ParseEventFn
from util.util import ParsePlayEventFn
from util.util import ParseArgs

# Defines the BigQuery schemas.
SESSION_SCHEMA = ('window_start:TIMESTAMP,' 'mean_duration:FLOAT')


class ComputeLatency(beam.DoFn):
    def __init__(self):
        super(ComputeLatency, self).__init__()
        self.dropped_sessions_no_events = Metrics.counter(
            self.__class__, 'dropped_sessions_no_events')
        self.dropped_sessions_too_many_events = Metrics.counter(
            self.__class__, 'dropped_sessions_too_many_events')
        self.dropped_sessions_no_play_events = Metrics.counter(
            self.__class__, 'dropped_sessions_no_play_events')

    def process(self, elem):
        _, vals = elem
        plays = vals['plays']
        events = vals['events']

        play_count = 0
        max_play_ts = 0
        for play in plays:
            play_count += 1
            max_play_ts = max(max_play_ts, long(play.timestamp))

        event_count = 0
        an_event = None
        for event in events:
            an_event = event
            event_count += 1

        if event_count == 0:
            self.dropped_sessions_no_events.inc()
        elif event_count > 1:
            self.dropped_sessions_too_many_events.inc()
        elif play_count == 0:
            self.dropped_sessions_no_play_events.inc()
        else:
            min_latency = long(an_event.timestamp) - max_play_ts
            yield (an_event.user, min_latency)


class DetectBadUsers(beam.DoFn):
    def process(self, elem, mean_latency=beam.DoFn.SideInputParam):
        user, latency = elem
        # Naive: compute bad users are users 5 times less than
        # the mean.
        if latency < mean / 5:
            yield user


def Run(argv=None):
    known_args, pipeline_args = ParseArgs(argv)
    pipeline_options = PipelineOptions(pipeline_args)
    pipeline_options.view_as(SetupOptions).save_main_session = True
    p = beam.Pipeline(options=pipeline_options)
    if known_args.topic:
        pipeline_options.view_as(StandardOptions).streaming = True

    project = pipeline_options.view_as(GoogleCloudOptions).project
    timestamp_attribute = 'timestamp_ms'
    events = None
    if (not known_args.topic or not known_args.play_topic):
        logging.fatal('topic and play_topic are required.')

    # [START EXERCISE 7]:
    # 1. Read game events with message id and timestamp.
    # 2. Parse events.
    events = (p
            | 'read_events' >> ChangeMe()
            | 'parse_events' >> ChangeMe()
            )

    # 1. Read play events with message id and timestamp.
    # 2. Parse events.
    play_events = (p
            | 'read_play_events' >> ChangeMe()
            | 'parse_play_events' >> ChangeMe()
            )

    # 1. Key events by event id.
    # 2. Sessionize.
    sessionized_events = (events
            | 'key_events_by_id' >> ChangeMe()
            | 'sessionize_events' >> ChangeMe()

    # 1. Key play events by event id.
    # 2. Sessionize.
    sessionized_plays = (play_events
            | 'key_plays_by_id' >> ChangeMe()
            | 'sessionize_plays' >> ChangeMe()

    # 1. Join events using CoGroupByKey
    # 2. Compute latency using ComputeLatency
    per_user_latency = (
            {'change':me, 'me':change}
            | 'cbk' >> ChangeMe()
            | 'compute_latency' >> ChangeMe()

    # 1. Get values of per user latencies
    # 2. Re-window into GlobalWindows that triggers repeatedly after 1000 new elements.
    # 3. Compute the global mean to be used as a side input.
    mean_latency = (per_user_latency
            | 'extract_latencies' >> ChangeMe()
            | 'global_window' >> ChangeMe()
            | 'compute_mean' >> ChangeMe()
            )
    # [END EXERCISE 7]

    # Filter out bad users.
    _ = (per_user_latency
            | 'detect_bad_users' >> beam.ParDo(
                DetectBadUsers(), mean_latency=mean_latency)
            | 'filter_duplicates' >> beam.WindowInto(
                window.GlobalWindows(), trigger=trigger.AfterCount(1),
                accumulation_mode=trigger.AccumulationMode.ACCUMULATING)
            | 'to_bq_schema' >> beam.Map(lambda x: {'user': x})
            | 'write_bad_users' >> beam.io.WriteToBigQuery(
                known_args.output_tablename, known_args.output_dataset, project, ('user:string'))
            )

    p.run().wait_until_finish()


if __name__ == '__main__':
    logging.getLogger().setLevel(logging.INFO)
    Run()
